s3/src/main/scala/akka/stream/alpakka/s3/auth/CanonicalRequest.scala

```diff
@@ -59,7 +59,13 @@ private[alpakka] object CanonicalRequest {
       pathEncodeRec(builder ++= uriEncodePath(head), tail)
   }
 
-  private def toHexUtf8(ch: Char): String = "%" + Integer.toHexString(ch.toInt)
+  private def toHexUtf8(ch: Char): String = ch match {
+    case asciiChar if ch >= 0 && ch <= 127 =>
+      "%" + Integer.toHexString(asciiChar.toInt).toUpperCase
+    case otherChar =>
+      // I guess this adds quite some overhead but the detailed api is not public
+      Path(otherChar.toString).toString()
```

Comment:
I suppose it would be neat to expose a `UriRendering.renderPath` that accepts a `keep: CharPredicate` instead of a `encodeFirstSegmentColons: Boolean`, that would make this a lot easier right?

I guess another approach might be to reuse the Path rendering and treat the `sub-delims` characters specially, like this:
```
private def pathEncode(path: Path): String =
    if (path.isEmpty) "/"
    else
      path.toString().flatMap {
        case ch if "!$&'()*+,;=".contains(ch) => "%" + Integer.toHexString(ch.toInt).toUpperCase
        case other => other.toString
      }
```